### ISSUE TITLE
New lint suggestion: Suspicious `{}` in string literal

### ISSUE THREAD

**Issue by emilk** [rust-lang/rust-clippy#11486]
### What it does

Look for curly braces in string literals and warn that the user may be missing a `format!` call.

### Advantage

I have several times forgotten to call `format!`, leading to subtle bugs. For instance:

```rs
on_error("There was an error: {error}");
```

### Drawbacks

There may be legitimate uses of curly-braces in strings, so there will likely be a lot of false positives. This should therefor be a _suspicious_ lint, that is opt-in.

### Example

```rust
on_error("There was an error: {error}");
```

Could be written as:

```rust
on_error(format!("There was an error: {error}"));
```

**Comment by y21** [comment]
cc #10195
It seems like it'd have FPs for JSON-like strings, but maybe we could try to parse the literal as a format string to see if it is one instead of only looking for just `{ .. }` (not sure if that ends up being too expensive) since most JSON is most likely not valid format syntax.

> This should therefor be a suspicious lint, that is opt-in.

I'm not sure if thats possible. Suspicious is warn by default and afaik needs to be opt out of manually by the user. Pedantic and restriction are allow by default (opt in)

**Comment by emilk** [comment]
You are right, this is a duplicate of https://github.com/rust-lang/rust-clippy/issues/10195 - I missed it when searching 🤦 

Parsing the `{contents}` would indeed be nice, or at least have a smart regex that checks the content is something like `\w*(:.*)?`

And yeah, I think I am confused about what suspicious means then :) pedantic/restriction is probably correct

**Comment by emilk** [comment]
Closing as duplicate of https://github.com/rust-lang/rust-clippy/issues/10195
